llvm/lib/DebugInfo/LogicalView/Readers/LVIRReader.cpp

```diff
@@ -0,0 +1,2309 @@
+// The following aditional steps need to be performed on them:
+// - If the lexical block doesn't have non-scope children, skip its
+//   emission and put its children directly to the parent scope.
+// The '--internal=id' is turned on just for debugging traces. Then
+// it is turned to its previous state.
+void LVIRReader::removeEmptyScopes() {
+  // Preserve current setting for '--internal=id'.
+  bool InternalID = options().getInternalID();
+
+  SmallVector<LVScope *> EmptyScopes;
+
+  // Delete lexically empty scopes.
+  auto DeleteEmptyScopes = [&]() {
+    if (EmptyScopes.empty())
+      return;
+
+    LLVM_DEBUG({
+      dbgs() << "\n** Collected empty scopes **\n";
+      for (auto Scope : EmptyScopes)
+        Scope->print(dbgs());
+    });
+
+    LVScope *Parent = nullptr;
+    for (auto Scope : EmptyScopes) {
+      Parent = Scope->getParentScope();
+      LLVM_DEBUG({
+        dbgs() << "Scope: " << Scope->getID() << ", ";
+        dbgs() << "Parent: " << Parent->getID() << "\n";
+      });
+
+      // If the target scope has lines, move them to the scope parent.
+      const LVLines *Lines = Scope->getLines();
+      if (Lines) {
+        LVLines Pack;
+        std::copy(Lines->begin(), Lines->end(), std::back_inserter(Pack));
+        for (LVLine *Line : Pack) {
+          if (Scope->removeElement(Line)) {
+            LLVM_DEBUG({ dbgs() << "Line: " << Line->getID() << "\n"; });
+            Line->resetParent();
+            Parent->addElement(Line);
+            Line->updateLevel(Parent, /*Moved=*/false);
+          }
+        }
+      }
+
+      if (Parent->removeElement(Scope)) {
+        const LVScopes *Scopes = Scope->getScopes();
+        if (Scopes) {
+          for (LVScope *Child : *Scopes) {
+            LLVM_DEBUG({ dbgs() << "Child: " << Child->getID() << "\n"; });
+            Child->resetParent();
+            Parent->addElement(Child);
+            Child->updateLevel(Parent, /*Moved=*/false);
+          }
+        }
+      }
+    }
+  };
+
+  // Traverse the scopes tree and collect those lexical blocks that does not
+  // have non-scope children. Do not include the lines as they are included
+  // in the logical view as a way to show their associated logical scope.
+  std::function<void(LVScope *)> TraverseScope = [&](LVScope *Current) {
+    auto IsEmpty = [](LVScope *Scope) -> bool {
+      return !Scope->getSymbols() && !Scope->getTypes() && !Scope->getRanges();
+    };
+
+    if (const LVScopes *Scopes = Current->getScopes()) {
+      for (LVScope *Scope : *Scopes) {
+        if (Scope->getIsLexicalBlock() && IsEmpty(Scope))
+          EmptyScopes.push_back(Scope);
+        TraverseScope(Scope);
+      }
+    }
+  };
+
+  TraverseScope(CompileUnit);
+  DeleteEmptyScopes();
+
+  // Restore setting for '--internal=id'.
+  if (InternalID)
+    options().setInternalID();
+}
+
+// During the IR Reader development, traverse all the logical elements
+// to check if they have been properly constructed (finalized).
```

Comment:
This comment is a bit confusing; maybe you meant
```suggestion
// During the IR-to-logical-view construction, traverse all the logical elements
// to check if they have been properly constructed (finalized).
```
?